### PR TITLE
Optional database encryption support on database creation.

### PR DESCRIPTION
**Pull request by xq262144**
Add several bugfixes including:
1. Optional database encryption support on database creation.
2. SQLChain layer block production race condition bugfix.
3. Client connection id allocation and error handling.
4. Deadly simple runtime space resource limit checker.
5. Merged new command line tool for SQL query & database creation/destroy.

### CODE REVIEW

**Review by leventeliu — APPROVED** [review]
LGTM.

File: cmd/idminer/main.go [hunk 1]
```diff
@@ -176,3 +211,88 @@ func runKeytool() {
...
+		os.Exit(-1)
+		return
+	}
+
+	// print the response
+	if resBytes, err := json.MarshalIndent(resp, "", "  "); err != nil {
+		log.Fatalf("marshal response failed: %v", err)
+		os.Exit(-1)
+		return
+	} else {
```
Comment (covenantdb) [inline cmd/idminer/main.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)

**Review by auxten — APPROVED** [review]
nice!

File: cmd/idminer/main.go (hunk 1 above)
Comment (covenantdb) [inline cmd/idminer/main.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)

**Pull request merged**